packages/web/utils/meTokens.tsx

```diff
@@ -0,0 +1,204 @@
+    case DAIADDRESS:
+      id = 'dai';
+      break;
+    default:
+      id = 'dai';
+  }
+  const tokenResponse = await fetch(`${COINGECKO_API_URL}${id}${COINQUERY}`);
+  const tokenJson = await tokenResponse.json();
+  return {
+    image: tokenJson.image.small,
+    currentPrice: tokenJson.market_data.current_price.usd,
+  };
+};
+
+const { mainnetRPC } = CONFIG;
+
+// Use mainnet provider for reads, to ensure anyone can view the profile regardless of Netork.
+const mainnetProvider = new ethers.providers.JsonRpcProvider(mainnetRPC);
+
+// Keeping it clean
+function getContractWithSigner(
+  contractAddress: string,
+  abi: any,
+  signer: Signer,
+) {
+  const contract = new ethers.Contract(contractAddress, abi, signer);
+  const contractWithSigner = contract.connect(signer);
+  return contractWithSigner;
+}
+
+// Read
+export const getMeTokenFor = async (ownerAddress: string) => {
+  if (!ownerAddress) return '';
+  const registry = getContractWithSigner(
+    metokenDiamond,
+    MeTokensRegistryABI,
+    mainnetProvider.getSigner(ownerAddress),
+  );
+  const meTokenAddress = await registry.getOwnerMeToken(ownerAddress);
+
+  return meTokenAddress;
+};
+
+export const getErc20TokenData = async (
+  tokenAddress: string,
+  owner: string,
+) => {
+  if (!owner) return '';
+  const erc20 = getContractWithSigner(
+    tokenAddress,
+    erc20Abi,
+    mainnetProvider.getSigner(owner),
+  );
+
+  const tokenData = {
+    symbol: (await erc20.symbol()) || '',
+    name: (await erc20.name()) || '',
+    balance: (await erc20.balanceOf(owner)) || '',
+  };
+
+  return { ...tokenData };
+};
+
+export const getMeTokenInfo = async (tokenAddress: string, owner: string) => {
+  if (!tokenAddress) return undefined;
+  const orbis = new Orbis();
+  const { data } = await orbis.getDids(owner);
+  const signer = mainnetProvider.getSigner(owner);
+
+  const erc20 = getContractWithSigner(tokenAddress, erc20Abi, signer);
+  const registry = getContractWithSigner(
+    metokenDiamond,
+    MeTokensRegistryABI,
+    signer,
+  );
+  const hub = getContractWithSigner(metokenDiamond, HubFacetAbi, signer);
+  const tokenInfo = await registry.getMeTokenInfo(tokenAddress);
+
+  const tokenData = {
+    symbol: `$${await erc20.symbol()}`,
+    profilePicture:
+      data[0]?.details.profile.pfp || 'https://tinyurl.com/mr29vhmk',
+    tokenAddress,
+    collateral: await hub
+      .getBasicHubInfo(tokenInfo?.hubId)
+      .then((res: { asset: string }) => res.asset),
+  };
+
+  return { ...tokenData };
+};
+
+export const preview = async (
+  meTokenAddress: string, // addr meToken
+  amount: BigNumber, // eg, 10e18 (CBOB)
+  senderAddress: string,
+  type: string,
+) => {
+  if (!amount || !meTokenAddress) return BigNumber.from(0);
+  const meTokenFoundry = await new Contract(
+    metokenDiamond,
+    FoundryFacetAbi,
+    mainnetProvider.getSigner(senderAddress),
+  );
+  if (type === 'mint') {
+    const tx = await meTokenFoundry
```

Comment:
probably shouldn't be using `await` and `.then`

Reply:
Resolved by using try/catch blocks, example:

```
   try {
      const tx = await meTokenFoundry
      .calculateMeTokensMinted(meTokenAddress, amount)
      return tx;
    } catch (e) {
      throw e;
    }
 
```